File: fifo_page.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <unordered_set>

using namespace std;

void FIFOPageReplacement(const vector<int>& pages, int capacity) {
    queue<int> pageQueue;
    unordered_set<int> pageSet;

    int pageFaults = 0;

    cout << "Page Reference\t" << endl;

    for (int page : pages) {
        cout << page << "\t";

        if (pageSet.find(page) == pageSet.end()) {
            if (pageQueue.size() == capacity) {
                int frontPage = pageQueue.front();
                pageQueue.pop();
                pageSet.erase(frontPage);
            }

            pageQueue.push(page);
            pageSet.insert(page);
            pageFaults++;

            queue<int> tempQueue = pageQueue;
            while (!tempQueue.empty()) {
                cout << tempQueue.front() << " ";
                tempQueue.pop();
            }
        }
        else {
            queue<int> tempQueue = pageQueue;
            while (!tempQueue.empty()) {
                cout << tempQueue.front() << " ";
                tempQueue.pop();
            }
        }

        cout << endl;
    }

    cout << "Total Page Faults: " << pageFaults << endl;
}

int main() {
    vector<int> pages;
    int capacity,pageNo,x;

    cout<<"Enter capacity";
    cin>>capacity;
    cout<<"Enter number of page reference:";
    cin>>pageNo;
    cout<<"Enter page references:";
    for (int i = 0; i < pageNo; i++)
    {
        
        cin>>x;
        pages.push_back(x);
    }

    FIFOPageReplacement(pages, capacity);

    return 0;
}

File: fifopage.py
from queue import Queue
def fifo_page(pages,frame):
    pageQueue = Queue()
    pagefault = 0
    print("Page\tReference")
    for page in pages:
        print(page,end="\t")
        if page not in list(pageQueue.queue):
            if len(list(pageQueue.queue)) == frame:
                pageQueue.get()
            pageQueue.put(page)
            pagefault +=1
            print(list(pageQueue.queue))                
        else:
            print(list(pageQueue.queue))
        print("")
    print(pagefault)
    
pages = [4,1,2,4,5]
frame = 3  
fifo_page(pages,frame)